pkg/apply/destroyer.go

```diff
@@ -0,0 +1,105 @@
+// It currently requires IOStreams, but this is a legacy from when
+// the ApplyOptions were responsible for printing progress. This is now
+// handled by a separate printer with the KubectlPrinterAdapter bridging
+// between the two.
+func NewDestroyer(factory util.Factory, ioStreams genericclioptions.IOStreams) *Destroyer {
+	return &Destroyer{
+		ApplyOptions: apply.NewApplyOptions(ioStreams),
+		PruneOptions: prune.NewPruneOptions(),
+		factory:      factory,
+		ioStreams:    ioStreams,
+	}
+}
+
+// Destroyer performs the step of grabbing all the previous inventory objects and
+// prune them. This also deletes all the previous inventory objects
+type Destroyer struct {
+	factory      util.Factory
+	ioStreams    genericclioptions.IOStreams
+	ApplyOptions *apply.ApplyOptions
+	PruneOptions *prune.PruneOptions
+
+	DryRun bool
+}
+
+// Initialize sets up the Destroyer for actually doing an destroy against
+// a cluster. This involves validating command line inputs and configuring
+// clients for communicating with the cluster.
+func (d *Destroyer) Initialize(cmd *cobra.Command) error {
+	err := d.ApplyOptions.Complete(d.factory, cmd)
+	if err != nil {
+		return errors.WrapPrefix(err, "error setting up ApplyOptions", 1)
+	}
+	err = d.PruneOptions.Initialize(d.factory, d.ApplyOptions.Namespace)
+	if err != nil {
+		return errors.WrapPrefix(err, "error setting up PruneOptions", 1)
+	}
+
+	// Propagate dry-run flags.
+	d.ApplyOptions.DryRun = d.DryRun
+	d.PruneOptions.DryRun = d.DryRun
+
+	if err != nil {
+		return errors.WrapPrefix(err, "error creating resolver", 1)
+	}
+	return nil
+}
+
+// Run performs the destroy step. This happens asynchronously
+// on progress and any errors are reported back on the event channel.
+func (d *Destroyer) Run() <-chan Event {
+	ch := make(chan Event)
+
+	go func() {
+		defer close(ch)
+		adapter := &KubectlPrinterAdapter{
+			ch: ch,
+		}
+		// The adapter is used to intercept what is meant to be printing
+		// in the PruneOptions, and instead turn those into events.
+		d.PruneOptions.ToPrinter = adapter.toPrinterFunc()
+
+		infos, _ := d.ApplyOptions.GetObjects()
+
+		err := d.PruneOptions.Prune(infos)
+		if err != nil {
+			// If we see an error here we just report it on the channel and then
+			// give up. Eventually we might be able to determine which errors
+			// are fatal and which might allow us to continue.
+			ch <- Event{
+				EventType: ErrorEventType,
+				ErrorEvent: ErrorEvent{
+					Err: errors.WrapPrefix(err, "error pruning resources", 1),
+				},
+			}
+			return
+		}
+	}()
+	return ch
+}
+
+// SetFlags configures the command line flags needed for destroy
+// This is a temporary solution as we should separate the configuration
+// of cobra flags from the Destroyer.
+func (d *Destroyer) SetFlags(cmd *cobra.Command) {
+	d.ApplyOptions.DeleteFlags.AddFlags(cmd)
+	d.ApplyOptions.RecordFlags.AddFlags(cmd)
+	d.ApplyOptions.PrintFlags.AddFlags(cmd)
```

Comment:
We don't want to set the print flags. Since we don't use the print functionality from Kubectl, we currently ignore them.